src/inspectors/package/RustPackageInspector.ts

```diff
@@ -0,0 +1,415 @@
+            dependencies: isOptStringArray,
+          },
+          value,
+        )
+      ) {
+        return (value as unknown) as CargoLockPackage;
+      }
+    }
+
+    throw new Error('Could not parse Cargo.lock package');
+  }
+
+  private static parseLock(value: unknown): CargoLock {
+    if (isRecord(value)) {
+      if (Array.isArray(value['package'])) {
+        const result: {
+          package: CargoLockPackage[];
+        } = {
+          package: [],
+        };
+
+        result.package.push(...value['package'].map((p) => RustPackageInspector.parseLockPackage(p)));
+
+        return result;
+      }
+    }
+
+    throw new Error('Could not parse Cargo.lock');
+  }
+
+  private static parseSemver(version: string | undefined): PackageVersion {
+    let result = undefined;
+    if (version !== undefined) {
+      result = PackageInspectorBase.semverToPackageVersion(version);
+    }
+
+    return (
+      result ?? {
+        value: '0.0.0',
+        major: '0',
+        minor: '0',
+        patch: '0',
+      }
+    );
+  }
+
+  private static extractPackages(manifest: CargoManifest, lockFile?: CargoLock): Package[] {
+    const parseDep = (type: DependencyType, dep: ManifestDependency) => {
+      const name = dep.package ?? dep.name;
+
+      const version = RustPackageInspector.parseSemver(dep.version);
+      const lockVersion =
+        lockFile === undefined ? version : RustPackageInspector.parseSemver(lockFile.package.find((p) => p.name === name)?.version);
+      return {
+        dependencyType: type,
+        name,
+        requestedVersion: version,
+        lockfileVersion: lockVersion,
+      };
+    };
+    const parseRuntimeDep = parseDep.bind(undefined, DependencyType.Runtime);
+    const parseDevDep = parseDep.bind(undefined, DependencyType.Dev);
+
+    const result: Package[] = [];
+
+    result.push(
+      ...manifest.dependencies.map(parseRuntimeDep),
+      ...manifest['dev-dependencies'].map(parseDevDep),
+      // ...manifest["build-dependencies"].map(), // TODO
+      ...Object.values(manifest.target ?? {}).flatMap((target) =>
+        target.dependencies.map(parseRuntimeDep).concat(
+          ...target['dev-dependencies'].map(parseDevDep),
+          // ...target["build-dependencies"].map() // TODO
+        ),
+      ),
+    );
+
+    return result;
+  }
+}
+
+export interface ManifestDependency {
+  name: string;
+  version?: string;
+  path?: string;
+  git?: string;
+  branch?: string;
+  package?: string;
+  optional?: boolean;
+  'default-features'?: boolean;
+  features?: string[];
+  registry?: string;
+}
+export interface BinaryInfo {
+  name: string;
+  path: string;
+}
+
+interface DependecySet {
+  dependencies: ReadonlyArray<ManifestDependency>;
+  'dev-dependencies': ReadonlyArray<ManifestDependency>;
+  'build-dependencies': ReadonlyArray<ManifestDependency>;
+}
+// https://doc.rust-lang.org/cargo/reference/manifest.html
+export interface CargoManifest extends DependecySet {
```

Comment:
I don't think it's necessary to have all the optional properties. All the packages in all the languages we support return the same object. Please, return also the same object. Keep our conventions.

Reply:
I'm not sure what you mean. The inspector doesn't actually return anything, instead it sets the inherited variable `packages` and fills it with the dependencies and the code does do this.

I did fill out all the properties according to the official Cargo documentation because it makes it easier for future implementors to build off of it if they'd like to detect properties like the `edition` or `license`.

However, I do realize that I have exposed the `cargoManifest` variable publicly on the inspector. I did this because I needed access to the properties of cargo manifest for the sake of the compoment detector and I believe practice checkers will need access to this file as well, as they would for example for checking rust edition or license as I mentioned above. I wasn't sure how to do it better as there is no precedence because all the existing practices either work with files outside the package management or with packages themselves. The checks I'm talking about are akin to looking for the `license` key in `package.json` or simillar, which isn't done anywhere in the existing codebase yet.

Reply:
Yes, but we apply [YAGNI principle](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it), meaning a programmer should not add functionality until deemed necessary. And adding the `license`, `edition`, or other is not necessary for now. So if you keep the same structure as other inspectors - do you still need the `cargoManifest` public?

Reply:
Yes, I need access to it in the component detector. Regardless of how it turns out with the 2 components return value issue I still need it to check for the existence of a `bin` entry in the manifest.